File: src/control_example.cpp
#include "navigation_node.hpp"

/**
 * @brief Exemplo simples de navegação reativa em C++.
 *
 * Esta classe implementa um algoritmo básico de navegação: vira em direção ao objetivo,
 * anda em linha reta se alinhado, e para se houver obstáculo à frente.
 *
 * Thresholds configuráveis:
 * - DIST_THRESHOLD: Distância para considerar objetivo atingido (0.1 m).
 * - ANGLE_THRESHOLD: Ângulo para considerar alinhado (0.1 rad ~ 5.7°).
 * - OBSTACLE_THRESHOLD: Distância para detectar obstáculo (0.5 m).
 */
class ControlExample : public NavigationNode
{
public:
  ControlExample() : NavigationNode("control_example_cpp") {}

private:
  /**
   * @brief Loop de controle executado a 20 Hz.
   *
   * Implementa a lógica de navegação: verifica objetivo, calcula erros,
   * decide entre virar, andar ou parar.
   */
  void control_loop() override
  {
    if (!has_goal()) {
      stop();
      return;
    }

    double dist_to_goal = distance_to_goal();
    double angle_err = angle_to_goal();
    double front_dist = get_front_distance();

    // Thresholds
    const double DIST_THRESHOLD = 0.1;     // metros
    const double ANGLE_THRESHOLD = 0.1;    // radianos (~5.7 graus)
    const double OBSTACLE_THRESHOLD = 0.5; // metros

    if (dist_to_goal < DIST_THRESHOLD) {
      // Objetivo alcançado
      clear_goal();
      stop();
      RCLCPP_INFO(this->get_logger(), "Objetivo alcançado!");
      return;
    }

    if (front_dist < OBSTACLE_THRESHOLD) {
      // Obstáculo à frente, parar
      stop();
      RCLCPP_WARN(this->get_logger(), "Obstáculo à frente, parando.");
      return;
    }

    if (std::abs(angle_err) > ANGLE_THRESHOLD) {
      // Virar em direção ao objetivo
      double w = (angle_err > 0) ? 0.5 : -0.5; // velocidade angular
      publish_velocity(0.0, w);
    } else {
      // Andar em linha reta
      publish_velocity(0.2, 0.0); // velocidade linear
    }
  }
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ControlExample>());
  rclcpp::shutdown();
  return 0;
}

File: scripts/control_example.py
#!/usr/bin/env python3
"""
Exemplo simples de controle de navegação.

Este script implementa um algoritmo básico: gira para alinhar com o objetivo,
depois anda em linha reta. Para se encontra obstáculo à frente ou se chega ao objetivo.
"""

import math
import rclpy
from phi_p3dx_navigation.main import NavigationNode


class ControlExample(NavigationNode):
    """
    Navegador de exemplo: gira para o goal, anda em frente, para se encontra obstáculo.
    """

    def __init__(self):
        super().__init__(node_name='control_example_navigator', timer_period=0.1)

        # Parâmetros do controle
        self.front_threshold = 0.5  # Distância para parar por obstáculo (m)
        self.angle_threshold = math.radians(5.0)  # Ângulo para considerar alinhado (graus)
        self.linear_speed = 0.2    # Velocidade linear (m/s)
        self.angular_speed = 0.1   # Velocidade angular (rad/s)

    def on_goal(self) -> None:
        """Chamado quando um novo goal é definido."""
        self.get_logger().info('Novo goal recebido! Iniciando controle simples.')

    def _control_loop(self) -> None:
        """Loop de controle"""
        if not self.has_goal():
            self.stop()
            return

        if not self.has_laser_data():
            self.stop()
            return
        
        # Checa se esta proximo do objetivo
        if self.distance_to_goal() < self.front_threshold:
            self.clear_goal()
            self.stop()
            return

        # Calcular ângulo para o objetivo
        angle_err = self.angle_to_goal()
        self.get_logger().info(f'Angulo para objetivo={math.degrees(angle_err):.1f}°')


        if abs(angle_err) > self.angle_threshold:
            # Girar para alinhar
            w = self.angular_speed 
            if angle_err > 0:
                w = self.angular_speed 
            else:
                w = -self.angular_speed
            self.publish_velocity(0.0, w)
            self.get_logger().info(f'Girando: vel angular={w:.1f}')
        else:

            # Verificar obstáculo à frente
            front_dist = self.get_front_distance()
            if front_dist < self.front_threshold:
                self.stop()
                self.get_logger().info('Obstáculo à frente: parando.')
                return


            # Andar em frente
            self.publish_velocity(self.linear_speed, 0.0)
            self.get_logger().info('Andando em frente.')


def main(args=None):
    rclpy.init(args=args)
    navigator = ControlExample()
    rclpy.spin(navigator)
    navigator.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()